teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+# TEP-0002: Enable Custom Tasks
+
+aka non-Pod Tasks, "duck-typed" Tasks 🦆
+
+Original Google Doc proposal, visible to members of tekton-dev@: https://docs.google.com/document/d/10nQSeIse7Ld4fLg4lhfgUmNKtewfaFNET3zlMdRnBuQ/edit
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Validation](#validation)
+  - [Status Reporting](#status-reporting)
+  - [Pipeline Integration](#pipeline-integration)
+  - [Initial Update Timeout](#initial-update-timeout)
+  - [Parameter Passing](#parameter-passing)
+  - [Result Reporting](#result-reporting)
+  - [Cancellation](#cancellation)
+  - [Timeout](#timeout)
+  - [CLI and Dashboard Support](#cli-and-dashboard-support)
+  - [Results API](#results-api)
+  - [Changes to Triggers](#changes-to-triggers)
+  - [User Stories (optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+Enable integrators to define new "custom" Task types as CRDs, which can be run
+by creating a new `Run` object (a CRD type which Tekton will define and
+own). Custom Task authors must also implement a reconciling controller which
```

Comment:
It seems like the main three patterns for extensibility in Kubernetes are:

- CRD/reconciling controller
- webhook
- "plugin" (aka bash scripts installed in cluster via daemonset)

Controller is the most robust, powerful one, but unfortunately still has the highest barrier to entry. I wonder how much of this could alternatively (additionally?) be done via webhook.

I'm imagining:

- A CRD type is defined by Tekton to hold task/webhook address info
- Tekton calls out to that webhook when a Run of that type is created
- The service backing that webhook does stuff

The connection could either be kept alive the entire time (GPRC-style), or periodically polled by Tekton for updates.

It would be awesome to be able to implement a CustomTask via a knative service.

Reply:
@dlorenc Why not both? 🌮 

I think the key thing is that with this, you can build that (possibly in a few ways!).  Probably one of the first things I'd want to play with is exactly what you say: a Custom Task that is driven by lightweight function executions.

Reply:
I agree with Matt, I think the idea sounds really cool and I think this gives us an avenue to experiment and iterate and get there.